server/src/test/java/org/opensearch/index/shard/SegmentReplicationIndexShardTests.java

```diff
@@ -52,6 +58,54 @@ public void testReplicationCheckpointNotNullForSegReb() throws IOException {
         closeShards(indexShard);
     }
 
+    public void testSegmentReplication_Index_Update_Delete() throws Exception {
+        String mappings = "{ \"" + MapperService.SINGLE_MAPPING_NAME + "\": { \"properties\": { \"foo\": { \"type\": \"keyword\"} }}}";
+        try (ReplicationGroup shards = createGroup(2, settings, mappings, new NRTReplicationEngineFactory())) {
+            shards.startAll();
+            final IndexShard primaryShard = shards.getPrimary();
+
+            final int numDocs = randomIntBetween(100, 200);
+            for (int i = 0; i < numDocs; i++) {
+                shards.index(new IndexRequest(index.getName()).id(String.valueOf(i)).source("{\"foo\": \"bar\"}", XContentType.JSON));
+            }
+
+            primaryShard.refresh("Test");
+            replicateSegments(primaryShard, shards.getReplicas());
+
+            shards.assertAllEqual(numDocs);
+
+            for (int i = 0; i < numDocs; i++) {
+                // randomly update docs.
+                if (randomBoolean()) {
+                    shards.index(
+                        new IndexRequest(index.getName()).id(String.valueOf(i)).source("{ \"foo\" : \"baz\" }", XContentType.JSON)
+                    );
+                }
+            }
+
+            primaryShard.refresh("Test");
+            replicateSegments(primaryShard, shards.getReplicas());
+            shards.assertAllEqual(numDocs);
+
+            final List<DocIdSeqNoAndSource> docs = getDocIdAndSeqNos(primaryShard);
+            for (IndexShard shard : shards.getReplicas()) {
+                assertEquals(getDocIdAndSeqNos(shard), docs);
+            }
+            for (int i = 0; i < numDocs; i++) {
+                // randomly delete.
+                if (randomBoolean()) {
+                    shards.delete(new DeleteRequest(index.getName()).id(String.valueOf(i)));
```

Comment:
nit: We can count number of docs deleted and assert `docsAfterDelete.size() == numDocs-deletedDocs`

Reply:
After update & Delete i'm asserting the equality of each doc id not just counts.